### PR TITLE
Handle Error when the Deposit Address Creation Fails - Crypto Flow

### PR DESCRIPTION
**Pull request by SudharakaP**
Resolve https://github.com/opencollective/opencollective/issues/4602

image

### CODE REVIEW

File: server/paymentProviders/thegivingblock/index.js [hunk 1]
```diff
@@ -109,11 +109,14 @@ export const processOrder = async order => {
...
   // create wallet address
-  const { depositAddress, pledgeId } = await createDepositAddress(account, {
+  const { depositAddress, pledgeId, errorMessage, meta } = await createDepositAddress(account, {
     organizationId: account.data.organizationId,
     pledgeAmount: order.data.customData.pledgeAmount,
     pledgeCurrency: order.data.customData.pledgeCurrency,
   });
+  if (errorMessage) {
+    throw new Error(`The Giving Block: ${errorMessage} ${meta.errorCode}`);
+  }
```
Comment (znarf) [inline server/paymentProviders/thegivingblock/index.js:119]: Let's log the error and fail with a generic error message for now. Similar to Stripe:
https://github.com/opencollective/opencollective-api/blob/main/server/paymentProviders/stripe/creditcard.js#L18 

We'll later be able to have a list of errors that we actually want to share with the end users:
https://github.com/opencollective/opencollective-api/blob/main/server/paymentProviders/stripe/creditcard.js#L344

**Review by znarf — CHANGES_REQUESTED** [review]

File: server/paymentProviders/thegivingblock/index.js (hunk 1 above)
Comment (SudharakaP) [reply server/paymentProviders/thegivingblock/index.js]: Makes sense. 👍🏽 Corrected this. Now it shows as below.  

image

File: server/paymentProviders/thegivingblock/index.js [hunk 2]
```diff
@@ -109,11 +112,15 @@ export const processOrder = async order => {
...
-  const { depositAddress, pledgeId } = await createDepositAddress(account, {
+  const { depositAddress, pledgeId, errorMessage, meta } = await createDepositAddress(account, {
     organizationId: account.data.organizationId,
     pledgeAmount: order.data.customData.pledgeAmount,
     pledgeCurrency: order.data.customData.pledgeCurrency,
   });
+  if (errorMessage) {
+    logger.error(`The Giving Block: ${errorMessage} ${meta.errorCode}`);
+    throw new Error(GENERIC_ERROR_MSG);
+  }
```
Comment (znarf) [inline server/paymentProviders/thegivingblock/index.js:123]: Why is this needed if we have `handleErrorsAndRetry`?

Add error handling there and refactor to avoid duplicate code.

Comment (SudharakaP) [reply server/paymentProviders/thegivingblock/index.js]: Yes, I've refactored code to handle the error within `handleErrorsAndRetry`. However it seems that there's cases where we could first have the `INVALID_JWT_TOKEN` error and then when we try to make the request again within the `handleErrorsAndRetry` then we could silently fail as we don't handle the response error there. I've handled this case as well. I could reproduce this in my local. 👍🏽 

Let me know how it looks now. 🤔

File: server/paymentProviders/thegivingblock/index.js [hunk 3]
```diff
@@ -46,13 +47,18 @@ async function handleErrorsAndRetry(result, path, options = {}, account = null)
...
+        if (result.data.errorMessage) {
+          throw new Error(`The Giving Block: ${result.data.errorMessage} ${result.data.meta.errorCode}`);
+        }
         return result.data;
       } catch (err) {
         error = err;
       }
-      throw error;
+      logger.error(error.message);
+      throw new Error(GENERIC_ERROR_MSG);
```
Comment (znarf) [inline server/paymentProviders/thegivingblock/index.js:58]: I don't see the point having this out of the `catch`.

Comment (SudharakaP) [reply server/paymentProviders/thegivingblock/index.js]: Yes makes sense. I've put this within the catch and simplified. Thanks. 👍🏽

**Review by znarf — APPROVED** [review]
